src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AwaitExpressionSpiller.cs

```diff
@@ -1156,11 +1156,18 @@ public override BoundNode VisitSequence(BoundSequence node)
             var value = VisitExpression(ref valueBuilder, node.Value);
 
             BoundSpillSequenceBuilder builder = null;
-            var sideEffects = VisitExpressionList(ref builder, node.SideEffects, forceSpill: valueBuilder != null, sideEffectsOnly: true);
+
+            // PROTOTYPE(patterns2): This does not properly handle side-effects that are statements. That occurs
+            // as a result of lowering a switch expression. Interesting cases to test include using `await`
+            // in a `when` clause of one of the switch expression arms, or in one of the result expressions.
+            // Until it is fixed, the cast below will fail whenever an async method contains a pattern switch.
+            var sideEffects = node.SideEffects.SelectAsArray(n => (BoundExpression)n);
+
+            sideEffects = VisitExpressionList(ref builder, sideEffects, forceSpill: valueBuilder != null, sideEffectsOnly: true);
 
             if (builder == null && valueBuilder == null)
             {
-                return node.Update(node.Locals, sideEffects, value, node.Type);
+                return node.Update(node.Locals, sideEffects.CastArray<BoundNode>(), value, node.Type);
```

Comment:
I think this is a safe cast so we can use `CastUp` instead of `CastArray`

Reply:
I intend to revert the changes to BoundSequence in a future PR, because we will need to use a lifting-lowering approach like we do for async for the switch expression (for a few reasons, including constraints around debugging and edit-and-continue, which require sequence points only occur where the stack is empty). I intend to introduce a new expression node with statement side-effects, which will be removed by a new lifting phase. Do please consider this kind of code soon obsolete.

---
In reply to: [175955824](https://github.com/dotnet/roslyn/pull/24059#discussion_r175955824) [](ancestors = 175955824)